src/v/cloud_storage/tests/cloud_storage_e2e_fuzz_test.cc

```diff
@@ -0,0 +1,336 @@
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/tests/manual_fixture.h"
+#include "cloud_storage/tests/produce_utils.h"
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "kafka/server/tests/produce_consume_utils.h"
+#include "model/fundamental.h"
+#include "storage/disk_log_impl.h"
+
+#include <seastar/core/io_priority_class.hh>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <iterator>
+
+static ss::logger fuzz_log("e2e_fuzz_test");
+namespace {
+const model::topic topic_name("tapioca");
+const model::partition_id pid(0);
+model::ntp ntp(model::kafka_namespace, topic_name, pid());
+
+// Like `% mod` but returns at least one.
+void mod_at_least_one(int* r, int mod) {
+    if (mod <= 1) {
+        mod = 2;
+    }
+    auto rn = *r;
+    *r = 1 + rn % (mod - 1);
+}
+
+enum class op_type {
+    min = 0,
+    produce = 0,     // (num_records, cardinality)
+    local_roll,      // (node_idx,    unused
+    local_gc,        // (node_idx,    num_segments)
+    local_gc_one,    // (node_idx,    unused)
+    local_compact,   // (node_idx,    unused)
+    upload_segments, // (unused,      unused)
+    upload_manifest, // (unused,      unused)
+    switch_leader,   // (node_idx,    unused)
+    max = switch_leader,
+};
+
+const char* op_names[]{
+  "op_type::produce",
+  "op_type::local_roll",
+  "op_type::local_gc",
+  "op_type::local_gc_one",
+  "op_type::local_compact",
+  "op_type::upload_segments",
+  "op_type::upload_manifest",
+  "op_type::switch_leader",
+};
+
+struct test_op {
+    op_type type{op_type::min};
+    int arg1{-1};
+    int arg2{-1};
+
+    ss::sstring to_string() const {
+        return ssx::sformat(
+          "{{ {}, {}, {} }}", op_names[static_cast<int>(type)], arg1, arg2);
+    }
+};
+
+const std::vector<std::pair<op_type, int>> op_weights{
+  {op_type::produce, 5},
+  {op_type::local_roll, 2},
+  {op_type::local_gc, 2},
+  {op_type::local_gc_one, 3},
+  {op_type::local_compact, 3},
+  {op_type::upload_segments, 3},
+  {op_type::upload_manifest, 5},
+  {op_type::switch_leader, 2},
+};
+
+test_op random_operation() {
+    static const int weight_total = [&] {
+        int total = 0;
+        for (const auto& [_, weight] : op_weights) {
+            total += weight;
+        }
+        return total;
+    }();
+    auto r = random_generators::get_int(0, weight_total);
+    int sum = 0;
+    op_type type{op_type::max};
+    for (const auto& [op, weight] : op_weights) {
+        sum += weight;
+        if (sum > r) {
+            type = op;
+            break;
+        }
+    }
+    auto arg1 = random_generators::get_int(0, 100);
+    auto arg2 = random_generators::get_int(0, 100);
+    return {type, arg1, arg2};
+}
```

Comment:
question: I'm not sure I understand the intuition behind how the random op is chosen. Could you explain the process a bit?

Reply:
Added some comments. TLDR, we're picking the op randomly based on weights, e.g. so produces tend to happen more so there's typically more substantial data, and then each op gets a couple random values to work with that they'll tweak based on op type.

The args generation could probably be put into the op_type switch directly, if that bit's confusing.